### ISSUE TITLE
Nesting binson objects when using ->put() does not work

### ISSUE THREAD

**Issue by DmSProject** [assaabloy-ppi/binson-php#11]
At first I tried to use your example with ->put([.....]) with a predefined structure, but that did not work when needing to put a binson object inside another object. It worked if I grabbed the bytes, but then it breaks in the key-app.
ex:
$binsonObj->put(["rand"=>$rand,"contact"=>$binsonObj2]) gives a fatal error on "wrong type"

**Comment by ppmag** [comment]
Sorry, lost this in another thread, thank you for posting it separately.

The reason of your issue is simple, there are no such thing (for this moment) as "Binson object" in my version of API.  Just BinsonWriter and BinsonParser.  

It looked for me redundant to create "binson objects", when we have powerful native PHP entities - associative arrays, which can itself (with few considerations), represent any simple or complex binson-object or binson-array.

So, $writer->put()  doesn't create "binson objects", it simply serialize associative array of random depth to binson raw bytestream, nothing more.

To overcome your issue, you just need to prepare your data in form of relevant PHP array(-s).
For example:

```php
$request = ['client_pubkey' => '7346464'];  // got from server

//building response, include whole original request in responce's 'req' field
$responce = ['dest'=>'client_1', 'action'=>'doit', 'req'=>$request];

$b = "";
$writer = new BinsonWriter($b);
$writer->put($responce);
echo bin2hex($b);
```
Above looks for me enough to cover all use-cases, but if don't, I'll redesign.
Right now thinking on it....
Please send me any counterexamples you have or will have during your integration process.

**Comment by DmSProject** [comment]
As of yesterday, before reading your replies, I started to think about the binson object as nested associative arrays sorted alphabetically by key, that made it much easier to grasp how to use it.

So, if I can create a full structure as nested associative arrays and throw the whole thing to ->put() and get a binson object back, that's very nice.

**Comment by ppmag** [comment]
Feel free to test it, just please pull latest updates from repo.

This ->put(), serializer, under the hood, should work with any structures, otherwise pls open an issue as a bug.

On the other hand, deserializer still don't fully work as expected, I'm trying to resolve that. So, parsing as iterative or manual walking the tree is still the only option for now.

**Comment by ppmag** [comment]
Serializer & deserializer now should work, also simple calls `binson_encode()` and `binson_decode()` may be the way to go in 99% of usage cases. See top README for details.
[Fixed in 1417e4dbbde74a0bcc3493028ee7e8a000f7ea7a]
